### PR TITLE
[All] Reduce calls to getValue

### PR DESCRIPTION
**Pull request by alxbilger**
Mainly for:
- MechanicalObject::position
- SparseGridRamificationTopology::hexahedra
- SparseGridRamificationTopology::position

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by alxbilger** [comment]
Note: collision pipeline must be refactored if we want to reduce the ridiculous high number of calls to `getValue`. Every time there is a check between two primitives, there is a `getValue` call to `MechanicalObject::position`, `proximity` and `alarmDistance`.

**Comment by alxbilger** [comment]
When running caduceus scene for 100 time steps:

| Calls to updateIfDirty()                  | Before   | After    |      |   |
|-------------------------------------------|----------|----------|------|---|
| MechanicalObject::position                | 2269573  | 515920   | -77% |   |
| SparseGridRamificationTopology::hexahedra | 646323   | 77430    | -88% |   |
| SparseGridRamificationTopology::position  | 644268   | 24473    | -96% |   |
| RegularGridTopology::n                    | 19981966 | 10347192 | -48% |   |

**Comment by alxbilger** [comment]
[ci-build][with-all-tests]

**Comment by damienmarchal** [comment]
> Note: collision pipeline must be refactored if we want to reduce the ridiculous high number of calls to `getValue`. Every time there is a check between two primitives, there is a `getValue` call to `MechanicalObject::position`, `proximity` and `alarmDistance`.

Will be a very funny task to do. 

EDIT: CTRL+F on 'getValue()[' shows +1000 hits.

**Comment by damienmarchal** [comment]
This PR makes me think that msg_info() << "message" is in fact something like:
```cpp
if(notMuted(Base*))
      info::queue << "message"
```
With notMuted implemented like that:
```cpp
Base::noMuted() 
{ 
     f_printLog().getValue().       
}
 ``` 
 
 So either we don't use msg_info in loops (which is a good idea by the way, but not easy to track and enforce in our code base) or we reduce the cost of notMuted by using internal caching of the state.

File: Sofa/Component/Topology/Container/Constant/src/sofa/component/topology/container/constant/MeshTopology.cpp [hunk 1]
```diff
@@ -533,10 +533,13 @@ void MeshTopology::init()
...
+
+    const auto& hexahedra = seqHexahedra.getValue();
+
     if (nbPoints==0)
     {
         // looking for upper topology
-        if (!seqHexahedra.getValue().empty())
+        if (!hexahedra.empty())
             m_upperElementType = core::topology::TopologyElementType::HEXAHEDRON;
         else if (!seqTetrahedra.getValue().empty())
```
Comment (fredroy) [inline Sofa/Component/Topology/Container/Constant/src/sofa/component/topology/container/constant/MeshTopology.cpp]: I saw in the code that it is the same .getValue() party with seqEdges, seqTrianges, seqTetra....
Will you do it as well ?

Comment (alxbilger) [reply Sofa/Component/Topology/Container/Constant/src/sofa/component/topology/container/constant/MeshTopology.cpp]: It's a good point. It's done.

File: Sofa/Component/Topology/Container/Constant/src/sofa/component/topology/container/constant/MeshTopology.cpp [hunk 2]
```diff
@@ -535,17 +535,21 @@ void MeshTopology::init()
     BaseMeshTopology::init();
 
     const auto& hexahedra = seqHexahedra.getValue();
```
Comment (damienmarchal) [inline Sofa/Component/Topology/Container/Constant/src/sofa/component/topology/container/constant/MeshTopology.cpp]: Shouldn't it be better to use an accessor there ?

**Comment by damienmarchal** [comment]
Conflict fixed... CI pass, I switch to ready... and let @fredroy merge it ;)

File: Sofa/Component/Topology/Container/Constant/src/sofa/component/topology/container/constant/MeshTopology.cpp (hunk 2 above)
Comment (alxbilger) [reply Sofa/Component/Topology/Container/Constant/src/sofa/component/topology/container/constant/MeshTopology.cpp]: I don't know. Is there a difference between `getValue` and `ReadAccessor`?

**Pull request merged**